ToopherDotNet/ToopherDotNet.cs

```diff
@@ -8,333 +9,775 @@
+								StreamReader reader = new StreamReader (stream, Encoding.UTF8);
+								error_message = reader.ReadToEnd ();
+							}
+
+							String statusLine = httpResp.StatusCode.ToString () + " : " + httpResp.StatusDescription;
+
+							if (String.IsNullOrEmpty (error_message)) {
+								throw new RequestError (statusLine);
+							} else {
+
+								try {
+									// Attempt to parse JSON response
+									var json = (JsonObject)SimpleJson.SimpleJson.DeserializeObject (error_message);
+									parseRequestError (json);
+								} catch (RequestError e) {
+									throw e;
+								} catch (Exception) {
+									throw new RequestError (statusLine + " : " + error_message);
+								}
+							}
+
+							throw new RequestError (error_message, wex);
+						}
 
 						try {
-							// Attempt to parse JSON response
-							var json = (JsonObject)SimpleJson.SimpleJson.DeserializeObject (error_message);
-							parseRequestError (json);
-						} catch (RequestError e) {
-							throw e;
-						} catch (Exception) {
-							throw new RequestError (statusLine + " : " + error_message);
+							return SimpleJson.SimpleJson.DeserializeObject (response);
+						} catch (Exception ex) {
+							throw new RequestError ("Could not parse response", ex);
 						}
 					}
+				}
 
-					throw new RequestError (error_message, wex);
+				public JsonObject get (string endpoint, NameValueCollection parameters = null)
+				{
+					return (JsonObject)request ("GET", endpoint, parameters);
 				}
-			
-				try {
-					return SimpleJson.SimpleJson.DeserializeObject (response);
-				} catch (Exception ex) {
-					throw new RequestError ("Could not parse response", ex);
+
+				public JsonArray getArray (string endpoint, NameValueCollection parameters = null)
+				{
+					return (JsonArray)request ("GET", endpoint, parameters);
+				}
+
+				public JsonObject post (string endpoint, NameValueCollection parameters = null)
+				{
+					return (JsonObject) request ("POST", endpoint, parameters);
+				}
+
+				private void parseRequestError (JsonObject err)
+				{
+					long errCode = (long)err["error_code"];
+					string errMessage = (string)err["error_message"];
+					if (errCode == UserDisabledError.ERROR_CODE) {
+						throw new UserDisabledError ();
+					} else if (errCode == UserUnknownError.ERROR_CODE) {
+						throw new UserUnknownError ();
+					} else if (errCode == TerminalUnknownError.ERROR_CODE) {
+						throw new TerminalUnknownError ();
+					} else {
+						if (errMessage.ToLower ().Contains ("pairing has been deactivated")
+							|| errMessage.ToLower ().Contains ("pairing has not been authorized")) {
+							throw new PairingDeactivatedError ();
+						} else {
+							throw new RequestError (errMessage);
+						}
+					}
 				}
 			}
+		}
+	}
+
+
+	public class Pairing
+	{
+		private IDictionary<string, Object> rawResponse;
+		private ToopherApi api;
+
+		public object this[string key]
+		{
+			get
```

Comment:
Alas, there's a build error if you add parens or a semi-colon. I've verified this is [valid syntax](https://msdn.microsoft.com/en-us/library/ms228503.aspx) for C#.  I did edit these to be on one-line, to be consistent with the other `get` and `set` in the file.